### PR TITLE
[RFC] Add runtime symbol information

### PR DESCRIPTION
**Pull request by lorc**
TLDR: We will get this:

> E/TC:1 0 Panic 'just testing' at core/tee/socket.c:57 <socket_open>
> E/TC:1 0 Call stack:
> E/TC:1 0  0x000000000e1096fc <print_kernel_stack+0x20>
> E/TC:1 0  0x000000000e1144bc <__do_panic+0x54>
> E/TC:1 0  0x000000000e125780 <socket_open+0x114>
> E/TC:1 0  0x000000000e10445c <pseudo_ta_enter_invoke_cmd+0xb0>
> E/TC:1 0  0x000000000e1156c0 <tee_ta_invoke_command+0xc4>
> E/TC:1 0  0x000000000e11b2b0 <syscall_invoke_ta_command+0xec>
> E/TC:1 0  0x000000000e10fb28 <tee_svc_do_call+0xd8>
> E/TC:1 0  0x000000000e102858 <el0_svc+0x68>

This is a fully-functional PoC that adds symbols information to TEE binary. It is loosely based on Linux kallsyms, but is written from scratch. Maybe, this was not the best idea, because python implementation is pretty slow: it takes about 2 seconds to process the symbols. Taking into account that we need two passes, it adds 4 seconds to the build time. Actually, there are room for improvement because even right now I see a number of ways to speed up things. But it is okay for PoC/RFC I think.

Speaking of overhead, it will add 4 seconds to the build time and about 30 kB to the binary size (checked for QEMU v8 build). Again, I see there some room for improvement, so I probably can shave a few kBs.

How it is working: at a high level we have two additional structures: symbols table, which is represented as sorted array of the following structures.

```
struct syms_table
{
	uint32_t offset; //<-- This is symbol offset from OP-TEE address space beginning  
	uint16_t name_offset;
	uint8_t name_len;
	uint8_t pad;
};
```
As array is sorted, it is quite easy to find needed symbol by its address with binary search.

Entry in symbols table references to a symbol name in a separate array. Compression algorithm is described in the `gen_syms.py` script.

To add symbols information into binary we need to perform two additional linking
passes. It is crucial to make sure that symbol names and their count
will not change between those passes. This is why `gen_syms.py` skips the symbols which it generates itself.

First pass is need to build symbols table and symbol names. But their
addressed will be invalid, because during the second pass linker
will move symbols to accommodate newly added symbols. This is why we need the second pass.

After second pass script will update symbols addresses and
we will be ready for the final pass. 

While I'm happy with C and Python code, I made a huge mess in the makefiles. This is why changes to build system are in the separate commit. I'll certainly will need help from someone who is more familiar with OP-TEE build system to integrate my changes nicelly.

### CODE REVIEW

**Comment by jforissier** [comment]
Hey Volodymyr,

> While I'm happy with C and Python code, I made a huge mess in the makefiles. 

:laughing: 

>This is why changes to build system are in the separate commit. I'll certainly will need help from someone who is more familiar with OP-TEE build system to integrate my changes nicelly.

Sure, I can take a look but not before Wednesday because I'm off tomorrow. The `gensrcs` mechanism should work fine for this purpose I think, in case you want to give it a try.

**Comment by lorc** [comment]
>  The gensrcs mechanism should work fine for this purpose I think, in case you want to give it a try.

Yes, I wanted to try it. But I not sure that it will work for me, because it should be done in strict order:

1. Link `tee.elf.0`
2. Generate `syms_data.c.0`
3. Compile `syms_data.o.0`
4. Link `tee.elf.1` (with `syms_data.o.0`)
5. Generate `syms_data.c.1`
6. Compile `syms_data.o.1`
7. Link final `tee.elf` with `syms_data.o.1`

I know, that it is `make`'s job to build targets in the correct order, but I'm not sure if `gensrcs` will work in my case.

**Comment by jforissier** [comment]
Forget about `gensrcs`, it's not needed because regular recipes are good enough. The only tricky part is to build the generated C files with all the C flags defined for the TEE core. Well in fact it's not that hard ;-) Please see commit 896f767b544e ("core: generate and link symbol table when CFG_CORE_SYMS=y") in my [syms](https://github.com/jforissier/optee_os/commits/syms) branch. You can simply replace your commit "syms: hack into build system" with that one.

What about the pager? It seems to me, the symbols and related functions need to be in the unpaged part? That is not very realistic due to size, so perhaps the symbol resolution should be disabled when `CFG_WITH_PAGER=y`?

Now for the minor things:
- Build warnings: they cause Shippable CI to fail since it builds with `CFG_WERROR=y`. The checkpatch errors are not hard to fix either.
- I think `scripts/symbolize.py` should be modified slightly to print the "function at source:line" message after the `<func+offset>` part (IMO it's good to keep the symbol + offset at least as a consistency check).

**Comment by lorc** [comment]
I addressed  other comments. And added verbose printing in `abort.c`. Maybe even too verbose :) 
If community thinks that it is too verbose, I can leave symbol printing only for "abort at address 0xXXXX" message.  

I don't know why IBART fails right now. 

And there are are couple of checkpatch warnings, that should be ignored. Also, I fixed makefile, because it failed with `CFG_CORE_SYMS=n`.

**Comment by jforissier** [comment]
> I addressed other comments. And added verbose printing in abort.c. Maybe even too verbose :)
If community thinks that it is too verbose, I can leave symbol printing only for "abort at address 0xXXXX" message.

Personally I don't find this really useful, in fact I rarely use the register values for anything ;-)

> I don't know why IBART fails right now.

Unrelated to this PR, don't worry. Proposed fix in #3304.

File: core/arch/arm/kernel/unwind_arm32.c [hunk 1]
```diff
@@ -435,10 +436,22 @@ void print_stack_arm32(int level, struct unwind_state_arm32 *state,
 		       vaddr_t exidx, size_t exidx_sz,
 		       vaddr_t stack, size_t stack_size)
 {
+#ifdef CFG_CORE_SYMS
```
Comment (jforissier) [inline core/arch/arm/kernel/unwind_arm32.c]: Minor: moving the #ifdef outside the function may be more legible, and same for arm64 below.

File: scripts/symbolize.py [hunk 2]
```diff
@@ -370,14 +371,17 @@ def write(self, line):
...
+                    just = 70
+                else:
+                    div = match.end('addr')
+                    just = 0
+                self._out.write(line[:div].ljust(just))
                 res = self.resolve(addr)
                 res = self.pretty_print_path(res)
                 self._out.write(' ' + res)
-                self._out.write(line[post:])
+                self._out.write(line[div:])
```
Comment (jforissier) [inline scripts/symbolize.py]: Personally I would not introduce any justification. The symbol names are already nicely aligned after the addresses as  `<sym+0x??>`, so it's good enough.

In fact I'd prefer we strip the function name returned by `addr2line` when we have a symbol + offset, because it is redundant information. But let's keep it while to double check that the runtime symbol resolution is correct; when this is sufficiently tested we can remove it IMO.

File: core/kernel/symbols.c [hunk 3]
```diff
@@ -0,0 +1,105 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
```
Comment (jforissier) [inline core/kernel/symbols.c]: C++-style comment for the SPDX line in C files, please (`// SPDX...`)

**Comment by lorc** [comment]
> Personally I don't find this really useful, in fact I rarely use the register values for anything ;-)

It had a few uses in my case. But yeah, probably, it is no worth for the generic case.

> Of course if the algorithm can be optimized we should do it in the end grinning

I tried to optimize the slowest function by using numpy. But effect was opposite. It be come 5 times slower :) So, I'm still considering approaches. Maybe it is worth to rewrite it in C... If you don't mind the idea of host compiler dependency.

File: scripts/symbolize.py (hunk 2 above)
Comment (lorc) [reply scripts/symbolize.py]: As for text justification, without it I m getting this output:
```
E/TC:0 0  0x000000000e10445c <pseudo_ta_enter_invoke_cmd+0xb0> pseudo_ta_enter_invoke_cmd at core/arch/arm/kernel/pseudo_ta.c:195
E/TC:0 0  0x000000000e115b4c <tee_ta_invoke_command+0xc4> tee_ta_invoke_command at core/kernel/tee_ta_manager.c:760
E/TC:0 0  0x000000000e10ffb4 <tee_svc_do_call+0xd8> tee_svc_do_call at core/arch/arm/tee/arch_svc_a64.S:141
E/TC:0 0  0x000000000e102858 <el0_svc+0x68> el0_svc at core/arch/arm/kernel/thread_a64.S:629
```
This is the reason, why I added justification in the first place. But I can remove it of course, no problem there.

> In fact I'd prefer we strip the function name returned by addr2line when we have a symbol + offset, because it is redundant information.

This can be useful to ensure that you are using correct symbols. On other hand, I'm not the user `symbolize.py` (I prefer gdb/JTAG), and I don't know if such problem ever existed. So it is up to you.

Comment (jforissier) [reply scripts/symbolize.py]: > As for text justification, without it I m getting this output:
> [...]

Yeah looks good to me :)

> This can be useful to ensure that you are using correct symbols. 

Good point. Let's keep the full addr2line output then.

**Comment by jforissier** [comment]
> If you don't mind the idea of host compiler dependency.

Well I don't mind on principle but I'd rather not introduce such a dependency only to shave 1.5 seconds on build time TBH ;-)

File: core/arch/arm/kernel/unwind_arm32.c (hunk 1 above)
Comment (lorc) [reply core/arch/arm/kernel/unwind_arm32.c]: just to clarify: you are proposing to have two variants of `print_stack_arm32()`: with and without symbols support.  Am I correct?

Comment (jforissier) [reply core/arch/arm/kernel/unwind_arm32.c]: Yes.

**Comment by lorc** [comment]
I addressed all the comments, but now it fails to compile with clang. Looks like `nm` produces some invalid addresses.  I have LLVM 8.0.1 on my PC and looks like it is not enough to build OP-TEE.

But I tried to use output from `llvm-nm` on gcc-generated ELF and it is working quite well. So, probably, there is some issue with clang itself...

**Comment by jforissier** [comment]
To use clang with OP-TEE you need to build it as shown in commit https://github.com/OP-TEE/optee_os/commit/98d863a5c0b27d79c76dbe9fdda73a35b7cf7405. It's quite easy.

File: core/arch/arm/kernel/unwind_arm32.c [hunk 4]
```diff
@@ -430,6 +431,25 @@ TEE_Result relocate_exidx(void *exidx, size_t exidx_sz, int32_t offset)
 }
 
 #if (TRACE_LEVEL > 0)
+#ifdef CFG_CORE_SYMS
+
+void print_stack_arm32(int level, struct unwind_state_arm32 *state,
+		       vaddr_t exidx, size_t exidx_sz,
+		       vaddr_t stack, size_t stack_size)
+{
+	char symbol[64];
```
Comment (etienne-lms) [inline core/arch/arm/kernel/unwind_arm32.c]: `= {};`

File: core/arch/arm/kernel/unwind_arm64.c [hunk 5]
```diff
@@ -65,6 +66,25 @@ bool unwind_stack_arm64(struct unwind_state_arm64 *frame,
 }
 
 #if (TRACE_LEVEL > 0)
+#ifdef CFG_CORE_SYMS
+
+void print_stack_arm64(int level, struct unwind_state_arm64 *state,
+		       vaddr_t stack, size_t stack_size)
+{
+	char symbol[64];
```
Comment (etienne-lms) [inline core/arch/arm/kernel/unwind_arm64.c]: init

File: core/arch/arm/kernel/abort.c [hunk 6]
```diff
@@ -123,6 +124,37 @@ static __maybe_unused const char *fault_to_str(uint32_t abort_type,
 	}
 }
 
+#ifdef CFG_CORE_SYMS
+/*
+ * Helper function this returns formated symbol name (if any) from a static
```
Comment (etienne-lms) [inline core/arch/arm/kernel/abort.c]: s/this/that/

File: core/arch/arm/kernel/abort.c [hunk 7]
```diff
@@ -123,6 +124,37 @@ static __maybe_unused const char *fault_to_str(uint32_t abort_type,
...
+ * storage. This is to ease up printing of symbols.
+ *
+ * Do not try to call this function multiple times for the same format
+ * string. I.e. do NOT do this:
+ * EMSG("%s %s", get_sym_name(v1, c), get_sym_name(v2, c));
+ */
+static __maybe_unused const char *get_sym_name(vaddr_t va, size_t core_pos)
+{
+	/* In case we'll have abort on several cores at the same time  */
+	static char sym_name[CFG_TEE_CORE_NB_CORE][64];
```
Comment (etienne-lms) [inline core/arch/arm/kernel/abort.c]: init `= {};`  and `= 0;` below.

Double space after `same time  */` in inline comment.

**Comment by etienne-lms** [comment]
here for my 2 cents

**Comment by jforissier** [comment]
It looks like Clang generates weak symbols with no value for the 6 symbols `syms_table`, `syms_count` etc. while GCC does not. Therefore the output of `llvm-nm` (or `nm`) has 2 entries (symbol type and name) instead of 3 (value, type and name). Since these symbols are ignored anyway, we can simply drop the lines that don't have 3 words like so:

```diff
diff --git a/scripts/gen_syms.py b/scripts/gen_syms.py
index 091e630226..831b7796e7 100755
--- a/scripts/gen_syms.py
+++ b/scripts/gen_syms.py
@@ -178,6 +178,7 @@ def emit_data(symbols, tok_table):
 def main():
     lines = sys.stdin.readlines()
     lines = itertools.filterfalse(lambda x: x == '', lines)
+    lines = itertools.filterfalse(lambda x: len(x.split(' ')) != 3, lines)
     symbols = map(make_symbol, lines)
     symbols = itertools.filterfalse(lambda x: x.name in IGNORE_SYMS, symbols)
     symbols = list(symbols)
```

File: core/kernel/symbols.c [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+ */
+
+#include <stdio.h>
+#include <symbols.h>
+#include <util.h>
+
+/* All extern entries are auto-generated during first and second linking pass*/
+
+/* Symbol table */
+extern const struct syms_table syms_table[] __weak;
```
[the remark this answers is not in the document]
Comment (lorc) [reply core/kernel/symbols.c]: Yes, because those symbols are defined in other translation unit.

File: core/kernel/symbols.c [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+		data++;
+	}
+
+	return ret;
+}
+
+int syms_format_name_w_offset(vaddr_t addr, char *buf, size_t buf_len)
+{
+	size_t offset;
```
[the remark this answers is not in the document]
Comment (lorc) [reply core/kernel/symbols.c]: I'm sorry, I don't get it. Could you please elaborate?

[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/kernel/symbols.c]: sorry. I meant initialize each local variable where defined. See https://optee.readthedocs.io/en/latest/general/coding_standards.html#coding-standards, section that starts with `Variables are initialized according to these general guidelines:`.

Here it would make:
```
	size_t offset = 0;
	int len = 0;
	const struct syms_table *sym = find_symbol(addr, &offset);

	(...)

[the remark this answers is not in the document]
Comment (lorc) [reply core/kernel/symbols.c]: Ah, I see. Looks like I somehow missed this addition to the coding style. Thank you for pointing out.

File: core/kernel/symbols.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/kernel/symbols.c]: what surprise me here is both `extern` and `__weak` on the same declaration.

[the remark this answers is not in the document]
Comment (lorc) [reply core/kernel/symbols.c]: Yes... `extern` is needed because this symbol is defined in `syms_data.c.[01]`. And we need `__weak`, because during first linking pass those symbols are undefined, so linker will fail with `undefined symbol` error. 

`syms_data.c.0` will be generated only after the first linker pass.

File: core/kernel/symbols.c [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+/* Basically, absolute address of the first symbol */
+extern const vaddr_t syms_base __weak;
+/* Indexes of symbols in the compressing table */
+extern const uint16_t syms_token_index[257] __weak;
+/* Compressing table */
+extern const char syms_token_table[] __weak;
+
+static const struct syms_table *find_symbol(vaddr_t addr, size_t *offset)
+{
+	unsigned int a = 0, b = syms_count;
```
Comment (etienne-lms) [inline core/kernel/symbols.c]: minor: split over 2 lines. (declare 1 variable per line)

**Comment by lorc** [comment]
@etienne-lms, thank you for review. I addressed all your comments. Sometimes, not in exact way you asked me, but it should be fine. 

Also I fixed that issue with the clang.

Sorry for not providing incremental commits. It got very messy at some point.

**Comment by jforissier** [comment]
I don't think that "gen_syms.py: emit comment with symbol name in generated file" should be a separate commit.

File: core/include/symbols.h [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+ * Copyright (c) 2019, EPAM Systems
+ */
+#ifndef SYMBOLS_H
+#define SYMBOLS_H
+
+#include <types_ext.h>
+
+/*
+ * This struct is exported only because it is used in auto-generated
+ * syms_data.c.0 an syms_data.c.1
```
Comment (etienne-lms) [inline core/include/symbols.h]: s/an/and/

File: core/include/symbols.h [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+ * Copyright (c) 2019, EPAM Systems
+ */
+#ifndef SYMBOLS_H
+#define SYMBOLS_H
+
+#include <types_ext.h>
+
+/* This struct is exported only because it is used in syms_data.c */
+struct syms_table {
+	/* Symbol offset from syms_base */
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/include/symbols.h]: discard this comment

File: core/kernel/symbols.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/kernel/symbols.c]: Checkpatch is not happy with `extern` in C source. Maybe declare them from `symbols.h`.

File: core/arch/arm/kernel/unwind_arm32.c [hunk 13]
```diff
@@ -430,6 +431,25 @@ TEE_Result relocate_exidx(void *exidx, size_t exidx_sz, int32_t offset)
...
+	do {
+		syms_format_name_w_offset(state->registers[PC], symbol,
+					  sizeof(symbol) - 1);
+		symbol[sizeof(symbol) - 1] = 0;
+		trace_printf_helper_raw(level, true, " 0x%08" PRIx32 " %s",
+					state->registers[PC], symbol);
+	} while (unwind_stack_arm32(state, exidx, exidx_sz, stack, stack_size));
+}
+
+#else  /* CFG_CORE_SYMS */
```
Comment (etienne-lms) [inline core/arch/arm/kernel/unwind_arm32.c]: style: remove 1 of the 2 spaces here.

File: core/arch/arm/kernel/unwind_arm32.c [hunk 14]
```diff
@@ -442,6 +462,8 @@ void print_stack_arm32(int level, struct unwind_state_arm32 *state,
 	} while (unwind_stack_arm32(state, exidx, exidx_sz, stack, stack_size));
 }
 
+#endif	/* CFG_CORE_SYMS */
```
Comment (etienne-lms) [inline core/arch/arm/kernel/unwind_arm32.c]: style: replace tabulation with a space character.
same comment in `unwind_arm64.c`.

File: core/arch/arm/kernel/unwind_arm32.c (hunk 13 above)
Comment (etienne-lms) [reply core/arch/arm/kernel/unwind_arm32.c]: same comment in `unwind_arm64.c`.

File: core/include/symbols.h (hunk 11 above)
Comment (jforissier) [inline core/include/symbols.h]: `syms_data.0.c` and `syms_data.1.c`

File: core/arch/arm/kernel/abort.c [hunk 15]
```diff
@@ -123,6 +124,37 @@ static __maybe_unused const char *fault_to_str(uint32_t abort_type,
...
+ * string. I.e. do NOT do this:
+ * EMSG("%s %s", get_sym_name(v1, c), get_sym_name(v2, c));
+ */
+static __maybe_unused const char *get_sym_name(vaddr_t va, size_t core_pos)
+{
+	/* In case we'll have abort on several cores at the same time */
+	static char sym_name[CFG_TEE_CORE_NB_CORE][64] = {};
+	int ret = 0;
+
+	sym_name[core_pos][0] = ' ';
```
Comment (etienne-lms) [inline core/arch/arm/kernel/abort.c]: very minor: looks strange that this function inserts a space before the target label.

File: core/arch/arm/kernel/abort.c [hunk 16]
```diff
@@ -123,6 +124,37 @@ static __maybe_unused const char *fault_to_str(uint32_t abort_type,
...
+	ret = syms_format_name_w_offset(va, sym_name[core_pos] + 1,
+					sizeof(sym_name[core_pos]) - 1);
```
Comment (etienne-lms) [inline core/arch/arm/kernel/abort.c]: Should be `sizeof(sym_name[core_pos]) - 2` since 1 char skipped (space char) and last is terminal `\0`?

**Comment by lorc** [comment]
> I don't think that "gen_syms.py: emit comment with symbol name in generated file" should be a separate commit.

I added it to debugging purposes, really. shippable randomly fails during building for imx6 platforms.  Strangelly, this is different imx6 platform every time. I can't reproduce this fail on my setup. So, I just wanted to see which symbol causes this fault.

**Comment by jenswi-linaro** [comment]
Now (introduced with ASLR) we have embedded boot data:
https://github.com/OP-TEE/optee_os/blob/490b9f949c894708f74b0c8d2e2924c4ffc30aae/core/arch/arm/include/kernel/generic_boot.h#L11-L40

This means that we can get rid of the second link stage and also store the symbol info in TA memory or even non-secure memory for debug purposes.

**Comment by lorc** [comment]
Yes, makes sense. 

Also, those patches lack support of ASLR.

Unfortunately, right now I have no time to finish this PR :(

**Comment by jbech-linaro** [comment]
> Unfortunately, right now I have no time to finish this PR :(

We can open it up again later on when you have time to dig into it, a closed PR doesn't mean closed forever.
